addons/account/tests/test_reconciliation.py

```diff
@@ -454,6 +456,136 @@ def test_revert_payment_and_reconcile(self):
+                exchange_move = exchange_reconcile.exchange_move_id
+                exchange_reconcile.unlink()
+                return exchange_move
+
+        # Testing exchange_move.date < today
+        for move_type in ('in_invoice', 'out_invoice'):
+            inv = create_invoice(move_type, '2016-06-01')
+            pay = create_payment(inv, '2020-08-01')
+            exchange_move = unreconcile_payment_and_get_exchange_move(pay)
+            reverted_exchange_move = get_reverted_exchange_move(exchange_move)
+            # exchange_move.date < today => reverted_exchange_move.date is last day of exchange_move.date month
+            self.assertEqual(
+                reverted_exchange_move.date,
+                fields.Date.from_string('2020-08-31'),
+                'reversed exchange move date %s for move type %s should be 2020-08-31'
+                % (reverted_exchange_move.date, move_type),
+            )
+
+        # Testing exchange_move.date < lockdate
+        for move_type in ('in_invoice', 'out_invoice'):
+            inv = create_invoice(move_type, '2016-06-01')
+            pay = create_payment(inv, '2020-08-01')
+            self.company.period_lock_date = '2020-09-01'
+            exchange_move = unreconcile_payment_and_get_exchange_move(pay)
+            reverted_exchange_move = get_reverted_exchange_move(exchange_move)
+            # exchange_move.date < lockdate => reverted_exchange_move.date is last day of month of day after lockdate
+            self.assertEqual(
+                reverted_exchange_move.date,
+                fields.Date.from_string('2020-09-30'),
+                'reversed exchange move date %s for move type %s should be 2020-09-30'
+                % (reverted_exchange_move.date, move_type),
+            )
+            self.company.period_lock_date = None
+
+        # Testing exchange_move.date = today
+        for move_type in ('in_invoice', 'out_invoice'):
+            inv = create_invoice(move_type, '2016-06-01')
+            pay = create_payment(inv, '2020-10-01')  # today
+            exchange_move = unreconcile_payment_and_get_exchange_move(pay)
+            reverted_exchange_move = get_reverted_exchange_move(exchange_move)
+            # exchange_move.date = today => reverted_exchange_move.date is today
+            self.assertEqual(
+                reverted_exchange_move.date,
+                fields.Date.from_string('2020-10-01'),
+                'reversed exchange move date %s for move type %s should be 2020-10-01'
+                % (reverted_exchange_move.date, move_type),
+            )
+
+        # Testing making invoice and payment for past dates & with period_lock_date > unreconcile_date
+        for move_type in ('in_invoice', 'out_invoice'):
+            inv = create_invoice(move_type, '2016-01-01', '2016-02-01')
+            pay = create_payment(inv, '2017-03-01', '2017-04-01')
+            self.company.period_lock_date = '2017-05-02'
```

Comment:
can be outside the loop. Also, it should be `self.env.company.period_lock_date` instead